File: sdk/runanywhere-swift/Sources/RunAnywhere/Capabilities/Voice/Strategies/VAD/SimpleEnergyVAD.swift
import Foundation
import AVFoundation
import Accelerate
import os

/// Simple energy-based Voice Activity Detection
/// Based on WhisperKit's EnergyVAD implementation but simplified for real-time audio processing
public class SimpleEnergyVAD: NSObject, VADService {
    // MARK: - Properties

    private let logger = Logger(subsystem: "com.runanywhere.sdk", category: "SimpleEnergyVAD")

    /// Energy threshold for voice activity detection (0.0 to 1.0)
    /// Values above this threshold indicate voice activity
    public var energyThreshold: Float = 0.022

    /// Sample rate of the audio (typically 16000 Hz)
    public let sampleRate: Int

    /// Length of each analysis frame in samples
    public let frameLengthSamples: Int

    /// Speech activity callback
    public var onSpeechActivity: ((SpeechActivityEvent) -> Void)?

    /// Optional callback for processed audio buffers
    public var onAudioBuffer: ((Data) -> Void)?

    // State tracking
    private var isActive = false
    private var isCurrentlySpeaking = false
    private var consecutiveSilentFrames = 0
    private var consecutiveVoiceFrames = 0

    // Hysteresis parameters to prevent rapid on/off switching
    private let voiceStartThreshold = 2  // frames of voice to start
    private let voiceEndThreshold = 5   // frames of silence to end (0.5 seconds at 100ms frames) - balanced for continuity

    // Calibration properties
    private var isCalibrating = false
    private var calibrationSamples: [Float] = []
    private var calibrationFrameCount = 0
    private let calibrationFramesNeeded = 20  // ~2 seconds at 100ms frames
    private var ambientNoiseLevel: Float = 0.0
    private var calibrationMultiplier: Float = 4.0  // Threshold = ambientNoise * multiplier - lower for better speech detection

    // Debug statistics
    private var recentEnergyValues: [Float] = []
    private let maxRecentValues = 50
    private var debugFrameCount = 0

    // MARK: - Initialization

    /// Initialize the VAD with specified parameters
    /// - Parameters:
    ///   - sampleRate: Audio sample rate (default: 16000)
    ///   - frameLength: Frame length in seconds (default: 0.1 = 100ms)
    ///   - energyThreshold: Energy threshold for voice detection (default: 0.022)
    public init(
        sampleRate: Int = 16000,
        frameLength: Float = 0.1,
        energyThreshold: Float = 0.022
    ) {
        self.sampleRate = sampleRate
        self.frameLengthSamples = Int(frameLength * Float(sampleRate))
        self.energyThreshold = energyThreshold
        super.init()

        logger.info("SimpleEnergyVAD initialized - sampleRate: \(sampleRate), frameLength: \(self.frameLengthSamples) samples, threshold: \(energyThreshold)")
    }

    // MARK: - VADService Protocol Implementation

    /// Initialize the VAD service
    public func initialize() async throws {
        start()
        // Start automatic calibration
        await startCalibration()
    }

    /// Current speech activity state
    public var isSpeechActive: Bool {
        return isCurrentlySpeaking
    }

    /// Frame length in seconds
    public var frameLength: Float {
        return Float(frameLengthSamples) / Float(sampleRate)
    }

    /// Reset the VAD state
    public func reset() {
        stop()
        isCurrentlySpeaking = false
        consecutiveSilentFrames = 0
        consecutiveVoiceFrames = 0
    }

    // MARK: - Public Methods

    /// Start voice activity detection
    public func start() {
        guard !isActive else { return }

        isActive = true
        isCurrentlySpeaking = false
        consecutiveSilentFrames = 0
        consecutiveVoiceFrames = 0

        logger.info("SimpleEnergyVAD started")
    }

    /// Stop voice activity detection
    public func stop() {
        guard isActive else { return }

        // If currently speaking, send end event
        if isCurrentlySpeaking {
            isCurrentlySpeaking = false
            logger.info("🎙️ VAD: SPEECH ENDED (stopped)")
            onSpeechActivity?(.ended)
        }

        isActive = false
        consecutiveSilentFrames = 0
        consecutiveVoiceFrames = 0

        logger.info("SimpleEnergyVAD stopped")
    }

    /// Process an audio buffer for voice activity detection
    /// - Parameter buffer: AVAudioPCMBuffer containing audio data
    public func processAudioBuffer(_ buffer: AVAudioPCMBuffer) {
        guard isActive else { return }

        // Convert buffer to float array
        let audioData = convertBufferToFloatArray(buffer)
        guard !audioData.isEmpty else { return }

        // Calculate energy of the entire buffer
        let energy = calculateAverageEnergy(of: audioData)

        // Update debug statistics
        updateDebugStatistics(energy: energy)

        // Handle calibration if active
        if isCalibrating {
            handleCalibrationFrame(energy: energy)
            return  // Don't process voice activity during calibration
        }

        let hasVoice = energy > energyThreshold

        // Enhanced logging with more context
        let energyStr = String(format: "%.6f", energy)
        let thresholdStr = String(format: "%.6f", energyThreshold)
        let percentAboveThreshold = ((energy - energyThreshold) / energyThreshold) * 100

        if debugFrameCount % 10 == 0 {  // Log every 10th frame to reduce noise
            let avgRecent = recentEnergyValues.isEmpty ? 0 : recentEnergyValues.reduce(0, +) / Float(recentEnergyValues.count)
            let maxRecent = recentEnergyValues.max() ?? 0
            let minRecent = recentEnergyValues.min() ?? 0

            logger.info("📊 VAD Stats - Current: \(energyStr) | Threshold: \(thresholdStr) | Voice: \(hasVoice ? "✅" : "❌") | %Above: \(String(format: "%.1f%%", percentAboveThreshold)) | Avg: \(String(format: "%.6f", avgRecent)) | Range: [\(String(format: "%.6f", minRecent))-\(String(format: "%.6f", maxRecent))]")
        }
        debugFrameCount += 1

        // Update state based on voice detection
        updateVoiceActivityState(hasVoice: hasVoice)

        // Call audio buffer callback if provided
        if let audioData = bufferToData(buffer) {
            onAudioBuffer?(audioData)
        }
    }

    /// Process a raw audio array for voice activity detection
    /// - Parameter audioData: Array of Float audio samples
    /// - Returns: Whether speech is detected in current frame
    @discardableResult
    public func processAudioData(_ audioData: [Float]) -> Bool {
        guard isActive else { return false }
        guard !audioData.isEmpty else { return false }

        // Calculate energy
        let energy = calculateAverageEnergy(of: audioData)

        // Update debug statistics
        updateDebugStatistics(energy: energy)

        // Handle calibration if active
        if isCalibrating {
            handleCalibrationFrame(energy: energy)
            return false  // Don't process voice activity during calibration
        }

        let hasVoice = energy > energyThreshold

        // Enhanced debug logging
        let ratio = energy / energyThreshold
        logger.debug("🎤 VAD: Energy=\(String(format: "%.6f", energy)) | Threshold=\(String(format: "%.6f", self.energyThreshold)) | Ratio=\(String(format: "%.2fx", ratio)) | Voice=\(hasVoice ? "YES✅" : "NO❌") | Ambient=\(String(format: "%.6f", self.ambientNoiseLevel))")

        // Update state
        updateVoiceActivityState(hasVoice: hasVoice)

        return hasVoice
    }

    // MARK: - Private Methods

    /// Calculate the RMS (Root Mean Square) energy of an audio signal
    /// - Parameter signal: Array of audio samples
    /// - Returns: RMS energy value
    private func calculateAverageEnergy(of signal: [Float]) -> Float {
        guard !signal.isEmpty else { return 0.0 }

        var rmsEnergy: Float = 0.0
        vDSP_rmsqv(signal, 1, &rmsEnergy, vDSP_Length(signal.count))
        return rmsEnergy
    }

    /// Update voice activity state with hysteresis to prevent rapid switching
    /// - Parameter hasVoice: Whether voice was detected in current frame
    private func updateVoiceActivityState(hasVoice: Bool) {
        if hasVoice {
            consecutiveVoiceFrames += 1
            consecutiveSilentFrames = 0

            // Start speaking if we have enough consecutive voice frames
            if !isCurrentlySpeaking && consecutiveVoiceFrames >= voiceStartThreshold {
                isCurrentlySpeaking = true
                logger.info("🎙️ VAD: SPEECH STARTED (energy above threshold for \(self.consecutiveVoiceFrames) frames)")
                DispatchQueue.main.async { [weak self] in
                    self?.onSpeechActivity?(.started)
                }
            }
        } else {
            consecutiveSilentFrames += 1
            consecutiveVoiceFrames = 0

            // Stop speaking if we have enough consecutive silent frames
            if isCurrentlySpeaking && consecutiveSilentFrames >= voiceEndThreshold {
                isCurrentlySpeaking = false
                logger.info("🎙️ VAD: SPEECH ENDED (silence for \(self.consecutiveSilentFrames) frames)")
                DispatchQueue.main.async { [weak self] in
                    self?.onSpeechActivity?(.ended)
                }
            }
        }
    }

    /// Convert AVAudioPCMBuffer to Float array
    /// - Parameter buffer: Audio buffer to convert
    /// - Returns: Array of Float audio samples
    private func convertBufferToFloatArray(_ buffer: AVAudioPCMBuffer) -> [Float] {
        guard let channelData = buffer.floatChannelData else { return [] }

        let frameLength = Int(buffer.frameLength)
        let samples = channelData.pointee

        return Array(UnsafeBufferPointer(start: samples, count: frameLength))
    }

    /// Convert audio buffer to Data for callback
    /// - Parameter buffer: Audio buffer to convert
    /// - Returns: Data representation of audio samples
    private func bufferToData(_ buffer: AVAudioPCMBuffer) -> Data? {
        guard let channelData = buffer.floatChannelData else { return nil }

        let channelDataValue = channelData.pointee
        let channelDataCount = Int(buffer.frameLength)

        let samples = Array(UnsafeBufferPointer<Float>(
            start: channelDataValue,
            count: channelDataCount
        ))

        return samples.withUnsafeBufferPointer { bufferPointer in
            Data(buffer: bufferPointer)
        }
    }

    // MARK: - Calibration Methods

    /// Start automatic calibration to determine ambient noise level
    public func startCalibration() async {
        logger.info("🎯 Starting VAD calibration - measuring ambient noise for \(Float(self.calibrationFramesNeeded) * self.frameLength) seconds...")

        isCalibrating = true
        calibrationSamples.removeAll()
        calibrationFrameCount = 0

        // Wait for calibration to complete (this is a simplified approach)
        // In production, you'd want a more sophisticated async approach
        let timeoutSeconds = Float(calibrationFramesNeeded) * frameLength + 2.0
        try? await Task.sleep(nanoseconds: UInt64(timeoutSeconds * 1_000_000_000))

        if isCalibrating {
            // Force complete calibration if still running
            completeCalibration()
        }
    }

    /// Handle a frame during calibration
    private func handleCalibrationFrame(energy: Float) {
        guard isCalibrating else { return }

        calibrationSamples.append(energy)
        calibrationFrameCount += 1

        logger.debug("📏 Calibration frame \(self.calibrationFrameCount)/\(self.calibrationFramesNeeded): energy=\(String(format: "%.6f", energy))")

        if calibrationFrameCount >= calibrationFramesNeeded {
            completeCalibration()
        }
    }

    /// Complete the calibration process
    private func completeCalibration() {
        guard isCalibrating, !calibrationSamples.isEmpty else { return }

        // Calculate statistics from calibration samples
        let sortedSamples = calibrationSamples.sorted()
        let mean = calibrationSamples.reduce(0, +) / Float(calibrationSamples.count)
        let median = sortedSamples[sortedSamples.count / 2]
        let percentile75 = sortedSamples[min(sortedSamples.count - 1, Int(Float(sortedSamples.count) * 0.75))]
        let percentile90 = sortedSamples[min(sortedSamples.count - 1, Int(Float(sortedSamples.count) * 0.90))]
        let max = sortedSamples.last ?? 0

        // Use 90th percentile as ambient noise level (robust to occasional spikes)
        ambientNoiseLevel = percentile90

        // Calculate dynamic threshold with better minimum
        let oldThreshold = energyThreshold
        // Ensure minimum threshold is high enough to avoid false positives
        // but low enough to detect actual speech
        let minimumThreshold: Float = 0.008  // Balanced for better speech detection
        let calculatedThreshold = ambientNoiseLevel * calibrationMultiplier

        // Apply threshold with sensible bounds
        energyThreshold = Swift.max(calculatedThreshold, minimumThreshold)

        // Cap at reasonable maximum
        if energyThreshold > 0.05 {
            energyThreshold = 0.05
            logger.warning("⚠️ Calibration detected high ambient noise. Capping threshold at 0.05")
        }

        logger.info("✅ VAD Calibration Complete:")
        logger.info("  📊 Statistics: Mean=\(String(format: "%.6f", mean)), Median=\(String(format: "%.6f", median))")
        logger.info("  📊 Percentiles: 75th=\(String(format: "%.6f", percentile75)), 90th=\(String(format: "%.6f", percentile90)), Max=\(String(format: "%.6f", max))")
        logger.info("  🎯 Ambient Noise Level: \(String(format: "%.6f", self.ambientNoiseLevel))")
        logger.info("  🔧 Threshold: \(String(format: "%.6f", oldThreshold)) → \(String(format: "%.6f", self.energyThreshold))")

        isCalibrating = false
        calibrationSamples.removeAll()
    }

    /// Manually set calibration parameters
    public func setCalibrationParameters(multiplier: Float = 4.0) {
        calibrationMultiplier = Swift.max(3.0, Swift.min(8.0, multiplier))  // Clamp between 3x and 8x for better speech detection
        logger.info("📝 Calibration multiplier set to \(self.calibrationMultiplier)x")
    }

    /// Get current VAD statistics for debugging
    public func getStatistics() -> (current: Float, threshold: Float, ambient: Float, recentAvg: Float, recentMax: Float) {
        let recent = recentEnergyValues.isEmpty ? 0 : recentEnergyValues.reduce(0, +) / Float(recentEnergyValues.count)
        let maxValue = recentEnergyValues.max() ?? 0
        let current = recentEnergyValues.last ?? 0

        return (current: current, threshold: energyThreshold, ambient: ambientNoiseLevel, recentAvg: recent, recentMax: maxValue)
    }

    // MARK: - Debug Helpers

    /// Update debug statistics with new energy value
    private func updateDebugStatistics(energy: Float) {
        recentEnergyValues.append(energy)
        if recentEnergyValues.count > maxRecentValues {
            recentEnergyValues.removeFirst()
        }
    }
}

File: sdk/runanywhere-flutter/lib/components/vad/simple_energy_vad.dart
import 'dart:async';
import 'dart:math' as math;
import '../../core/module_registry.dart';
import '../../foundation/logging/sdk_logger.dart';

/// Speech activity events
enum SpeechActivityEvent {
  started,
  ended;

  String get value {
    switch (this) {
      case SpeechActivityEvent.started:
        return 'started';
      case SpeechActivityEvent.ended:
        return 'ended';
    }
  }
}

/// Simple energy-based Voice Activity Detection
/// Based on iOS WhisperKit's EnergyVAD implementation but simplified for real-time audio processing
class SimpleEnergyVAD implements VADService {
  final SDKLogger _logger = SDKLogger(category: 'SimpleEnergyVAD');

  /// Energy threshold for voice activity detection (0.0 to 1.0)
  /// Values above this threshold indicate voice activity
  double energyThreshold = 0.005; // Even lower threshold for better short phrase detection

  /// Base threshold before any adjustments
  double _baseEnergyThreshold = 0.005;

  /// Multiplier applied during TTS playback to prevent feedback
  double _ttsThresholdMultiplier = 3.0;

  /// Sample rate of the audio (typically 16000 Hz)
  final int sampleRate;

  /// Length of each analysis frame in samples
  final int frameLengthSamples;

  /// Frame length in seconds
  double get frameLength => frameLengthSamples / sampleRate;

  /// Speech activity callback
  void Function(SpeechActivityEvent)? onSpeechActivity;

  /// Optional callback for processed audio buffers
  void Function(List<int>)? onAudioBuffer;

  // State tracking
  bool _isActive = false;
  bool _isCurrentlySpeaking = false;
  int _consecutiveSilentFrames = 0;
  int _consecutiveVoiceFrames = 0;
  bool _isPaused = false; // Track paused state
  bool _isTTSActive = false; // Track if TTS is currently playing

  // Hysteresis parameters to prevent rapid on/off switching
  final int _voiceStartThreshold = 1; // frames of voice to start - reduced to 1 frame for better short phrase detection
  final int _voiceEndThreshold = 8; // frames of silence to end (0.8 seconds at 100ms frames) - shorter for quicker responsiveness

  // Enhanced hysteresis for TTS mode
  final int _ttsVoiceStartThreshold = 10; // Much more frames needed during TTS to prevent feedback
  final int _ttsVoiceEndThreshold = 5; // Quicker end during TTS

  // Calibration properties
  bool _isCalibrating = false;
  final List<double> _calibrationSamples = [];
  int _calibrationFrameCount = 0;
  final int _calibrationFramesNeeded = 20; // ~2 seconds at 100ms frames
  double _ambientNoiseLevel = 0.0;
  double _calibrationMultiplier = 2.5; // Threshold = ambientNoise * multiplier - higher to reduce false positives

  // Debug statistics
  final List<double> _recentEnergyValues = [];
  final int _maxRecentValues = 50;
  int _debugFrameCount = 0;

  /// Initialize the VAD with specified parameters
  SimpleEnergyVAD({
    this.sampleRate = 16000,
    double frameLength = 0.1,
    double energyThreshold = 0.005,
  })  : frameLengthSamples = (frameLength * sampleRate).toInt(),
        energyThreshold = energyThreshold {
    _logger.info(
      'SimpleEnergyVAD initialized - sampleRate: $sampleRate, frameLength: $frameLengthSamples samples, threshold: $energyThreshold',
    );
  }

  // VADService interface implementation
  @override
  Future<void> initialize({String? modelPath}) async {
    start();
    // Start automatic calibration
    await startCalibration();
  }

  @override
  bool get isReady => _isActive;

  @override
  Future<VADResult> detect({required List<int> audioData}) async {
    processAudioBuffer(audioData);
    return VADResult(
      hasSpeech: _isCurrentlySpeaking,
      confidence: energyThreshold,
    );
  }

  @override
  Future<void> cleanup() async {
    stop();
    _recentEnergyValues.clear();
    _calibrationSamples.clear();
  }

  /// Current speech activity state
  bool get isSpeechActive => _isCurrentlySpeaking;

  /// Reset the VAD state
  void reset() {
    stop();
    _isCurrentlySpeaking = false;
    _consecutiveSilentFrames = 0;
    _consecutiveVoiceFrames = 0;
  }

  /// Start voice activity detection
  void start() {
    if (_isActive) return;

    _isActive = true;
    _isCurrentlySpeaking = false;
    _consecutiveSilentFrames = 0;
    _consecutiveVoiceFrames = 0;

    _logger.info('SimpleEnergyVAD started');
  }

  /// Stop voice activity detection
  void stop() {
    if (!_isActive) return;

    // If currently speaking, send end event
    if (_isCurrentlySpeaking) {
      _isCurrentlySpeaking = false;
      _logger.info('🎙️ VAD: SPEECH ENDED (stopped)');
      onSpeechActivity?.call(SpeechActivityEvent.ended);
    }

    _isActive = false;
    _consecutiveSilentFrames = 0;
    _consecutiveVoiceFrames = 0;

    _logger.info('SimpleEnergyVAD stopped');
  }

  /// Process an audio buffer for voice activity detection
  void processAudioBuffer(List<int> buffer) {
    if (!_isActive) return;

    // Complete audio blocking during TTS - don't process at all
    if (_isTTSActive) {
      return;
    }

    if (_isPaused) {
      return;
    }

    if (buffer.isEmpty) return;

    // Convert 16-bit PCM samples to Float32
    final audioData = _convertPCMToFloat(buffer);

    // Calculate energy of the entire buffer
    final energy = _calculateAverageEnergy(audioData);

    // Update debug statistics
    _updateDebugStatistics(energy);

    // Handle calibration if active
    if (_isCalibrating) {
      _handleCalibrationFrame(energy);
      return; // Don't process voice activity during calibration
    }

    final hasVoice = energy > energyThreshold;

    // Enhanced logging with more context
    final percentAboveThreshold =
        ((energy - energyThreshold) / energyThreshold) * 100;

    if (_debugFrameCount % 10 == 0) {
      // Log every 10th frame to reduce noise
      final avgRecent = _recentEnergyValues.isEmpty
          ? 0.0
          : _recentEnergyValues.reduce((a, b) => a + b) /
              _recentEnergyValues.length;
      final maxRecent = _recentEnergyValues.isEmpty
          ? 0.0
          : _recentEnergyValues.reduce(math.max);
      final minRecent = _recentEnergyValues.isEmpty
          ? 0.0
          : _recentEnergyValues.reduce(math.min);

      _logger.info(
        '📊 VAD Stats - Current: ${energy.toStringAsFixed(6)} | '
        'Threshold: ${energyThreshold.toStringAsFixed(6)} | '
        'Voice: ${hasVoice ? "✅" : "❌"} | '
        '%Above: ${percentAboveThreshold.toStringAsFixed(1)}% | '
        'Avg: ${avgRecent.toStringAsFixed(6)} | '
        'Range: [${minRecent.toStringAsFixed(6)}-${maxRecent.toStringAsFixed(6)}]',
      );
    }
    _debugFrameCount++;

    // Update state based on voice detection
    _updateVoiceActivityState(hasVoice);

    // Call audio buffer callback if provided
    onAudioBuffer?.call(buffer);
  }

  /// Process a raw audio array for voice activity detection
  bool processAudioData(List<double> audioData) {
    if (!_isActive) return false;

    // Complete audio blocking during TTS - don't process at all
    if (_isTTSActive) {
      return false;
    }

    if (_isPaused) return false;
    if (audioData.isEmpty) return false;

    // Calculate energy
    final energy = _calculateAverageEnergy(audioData);

    // Update debug statistics
    _updateDebugStatistics(energy);

    // Handle calibration if active
    if (_isCalibrating) {
      _handleCalibrationFrame(energy);
      return false; // Don't process voice activity during calibration
    }

    final hasVoice = energy > energyThreshold;

    // Enhanced debug logging
    final ratio = energy / energyThreshold;
    _logger.debug(
      '🎤 VAD: Energy=${energy.toStringAsFixed(6)} | '
      'Threshold=${energyThreshold.toStringAsFixed(6)} | '
      'Ratio=${ratio.toStringAsFixed(2)}x | '
      'Voice=${hasVoice ? "YES✅" : "NO❌"} | '
      'Ambient=${_ambientNoiseLevel.toStringAsFixed(6)}',
    );

    // Update state
    _updateVoiceActivityState(hasVoice);

    return hasVoice;
  }

  /// Calculate the RMS (Root Mean Square) energy of an audio signal
  double _calculateAverageEnergy(List<double> signal) {
    if (signal.isEmpty) return 0.0;

    double sumSquares = 0.0;
    for (final sample in signal) {
      sumSquares += sample * sample;
    }

    return math.sqrt(sumSquares / signal.length);
  }

  /// Update voice activity state with hysteresis to prevent rapid switching
  void _updateVoiceActivityState(bool hasVoice) {
    // Use different thresholds based on TTS state
    final startThreshold =
        _isTTSActive ? _ttsVoiceStartThreshold : _voiceStartThreshold;
    final endThreshold =
        _isTTSActive ? _ttsVoiceEndThreshold : _voiceEndThreshold;

    if (hasVoice) {
      _consecutiveVoiceFrames++;
      _consecutiveSilentFrames = 0;

      // Start speaking if we have enough consecutive voice frames
      if (!_isCurrentlySpeaking &&
          _consecutiveVoiceFrames >= startThreshold) {
        // Extra validation during TTS to prevent false positives
        if (_isTTSActive) {
          _logger.warning(
            '⚠️ Voice detected during TTS playback - likely feedback! Ignoring.',
          );
          return;
        }

        _isCurrentlySpeaking = true;
        _logger.info(
          '🎙️ VAD: SPEECH STARTED (energy above threshold for $_consecutiveVoiceFrames frames)',
        );
        onSpeechActivity?.call(SpeechActivityEvent.started);
      }
    } else {
      _consecutiveSilentFrames++;
      _consecutiveVoiceFrames = 0;

      // Stop speaking if we have enough consecutive silent frames
      if (_isCurrentlySpeaking && _consecutiveSilentFrames >= endThreshold) {
        _isCurrentlySpeaking = false;
        _logger.info(
          '🎙️ VAD: SPEECH ENDED (silence for $_consecutiveSilentFrames frames)',
        );
        onSpeechActivity?.call(SpeechActivityEvent.ended);
      }
    }
  }

  /// Convert 16-bit PCM samples to Float32 (-1.0 to 1.0)
  List<double> _convertPCMToFloat(List<int> pcmSamples) {
    final floatSamples = <double>[];
    for (final sample in pcmSamples) {
      // Convert from 16-bit int (-32768 to 32767) to float (-1.0 to 1.0)
      floatSamples.add(sample / 32768.0);
    }
    return floatSamples;
  }

  // MARK: - Calibration Methods

  /// Start automatic calibration to determine ambient noise level
  Future<void> startCalibration() async {
    _logger.info(
      '🎯 Starting VAD calibration - measuring ambient noise for ${_calibrationFramesNeeded * frameLength} seconds...',
    );

    _isCalibrating = true;
    _calibrationSamples.clear();
    _calibrationFrameCount = 0;

    // Wait for calibration to complete
    final timeoutSeconds = _calibrationFramesNeeded * frameLength + 2.0;
    await Future.delayed(Duration(milliseconds: (timeoutSeconds * 1000).toInt()));

    if (_isCalibrating) {
      // Force complete calibration if still running
      _completeCalibration();
    }
  }

  /// Handle a frame during calibration
  void _handleCalibrationFrame(double energy) {
    if (!_isCalibrating) return;

    _calibrationSamples.add(energy);
    _calibrationFrameCount++;

    _logger.debug(
      '📏 Calibration frame $_calibrationFrameCount/$_calibrationFramesNeeded: energy=${energy.toStringAsFixed(6)}',
    );

    if (_calibrationFrameCount >= _calibrationFramesNeeded) {
      _completeCalibration();
    }
  }

  /// Complete the calibration process
  void _completeCalibration() {
    if (!_isCalibrating || _calibrationSamples.isEmpty) return;

    // Calculate statistics from calibration samples
    final sortedSamples = List<double>.from(_calibrationSamples)..sort();
    final mean =
        _calibrationSamples.reduce((a, b) => a + b) / _calibrationSamples.length;
    final median = sortedSamples[sortedSamples.length ~/ 2];
    final percentile75 = sortedSamples[
        math.min(sortedSamples.length - 1, (sortedSamples.length * 0.75).toInt())];
    final percentile90 = sortedSamples[
        math.min(sortedSamples.length - 1, (sortedSamples.length * 0.90).toInt())];
    final max = sortedSamples.last;

    // Use 90th percentile as ambient noise level (robust to occasional spikes)
    _ambientNoiseLevel = percentile90;

    // Calculate dynamic threshold with better minimum
    final oldThreshold = energyThreshold;
    // Ensure minimum threshold is high enough to avoid false positives
    // but low enough to detect actual speech
    final minimumThreshold =
        math.max(_ambientNoiseLevel * 2.5, 0.006); // At least 2.5x ambient or 0.006
    final calculatedThreshold = _ambientNoiseLevel * _calibrationMultiplier;

    // Apply threshold with sensible bounds
    energyThreshold = math.max(calculatedThreshold, minimumThreshold);

    // Cap at reasonable maximum - balanced for speech detection without false positives
    if (energyThreshold > 0.020) {
      energyThreshold = 0.020;
      _logger.warning(
        '⚠️ Calibration detected high ambient noise. Capping threshold at 0.020',
      );
    }

    _logger.info('✅ VAD Calibration Complete:');
    _logger.info(
      '  📊 Statistics: Mean=${mean.toStringAsFixed(6)}, Median=${median.toStringAsFixed(6)}',
    );
    _logger.info(
      '  📊 Percentiles: 75th=${percentile75.toStringAsFixed(6)}, 90th=${percentile90.toStringAsFixed(6)}, Max=${max.toStringAsFixed(6)}',
    );
    _logger.info(
      '  🎯 Ambient Noise Level: ${_ambientNoiseLevel.toStringAsFixed(6)}',
    );
    _logger.info(
      '  🔧 Threshold: ${oldThreshold.toStringAsFixed(6)} → ${energyThreshold.toStringAsFixed(6)}',
    );

    _isCalibrating = false;
    _calibrationSamples.clear();
  }

  /// Manually set calibration parameters
  void setCalibrationParameters({double multiplier = 2.5}) {
    _calibrationMultiplier =
        math.max(2.0, math.min(4.0, multiplier)); // Clamp between 2.0x and 4.0x
    _logger.info('📝 Calibration multiplier set to ${_calibrationMultiplier}x');
  }

  /// Get current VAD statistics for debugging
  Map<String, double> getStatistics() {
    final recent = _recentEnergyValues.isEmpty
        ? 0.0
        : _recentEnergyValues.reduce((a, b) => a + b) / _recentEnergyValues.length;
    final maxValue =
        _recentEnergyValues.isEmpty ? 0.0 : _recentEnergyValues.reduce(math.max);
    final current = _recentEnergyValues.isEmpty ? 0.0 : _recentEnergyValues.last;

    return {
      'current': current,
      'threshold': energyThreshold,
      'ambient': _ambientNoiseLevel,
      'recentAvg': recent,
      'recentMax': maxValue,
    };
  }

  // MARK: - Pause and Resume

  /// Pause VAD processing
  void pause() {
    if (_isPaused) return;
    _isPaused = true;
    _logger.info('⏸️ VAD paused');

    // If currently speaking, send end event
    if (_isCurrentlySpeaking) {
      _isCurrentlySpeaking = false;
      onSpeechActivity?.call(SpeechActivityEvent.ended);
    }

    // Clear recent energy values to avoid false positives when resuming
    _recentEnergyValues.clear();
    _consecutiveSilentFrames = 0;
    _consecutiveVoiceFrames = 0;
  }

  /// Resume VAD processing
  void resume() {
    if (!_isPaused) return;

    _isPaused = false;

    // Reset state for clean resumption
    _isCurrentlySpeaking = false;
    _consecutiveSilentFrames = 0;
    _consecutiveVoiceFrames = 0;
    // Clear any accumulated energy values to start fresh
    _recentEnergyValues.clear();
    _debugFrameCount = 0;

    _logger.info('▶️ VAD resumed');
  }

  // MARK: - TTS Feedback Prevention

  /// Notify VAD that TTS is about to start playing
  void notifyTTSWillStart() {
    _isTTSActive = true;

    // Save base threshold
    _baseEnergyThreshold = energyThreshold;

    // Increase threshold significantly to prevent TTS audio from triggering VAD
    final newThreshold = energyThreshold * _ttsThresholdMultiplier;
    energyThreshold = math.min(newThreshold, 0.1); // Cap at 0.1 to prevent complete deafness

    _logger.info(
      '🔊 TTS starting - VAD completely blocked and threshold increased from ${_baseEnergyThreshold.toStringAsFixed(6)} to ${energyThreshold.toStringAsFixed(6)}',
    );

    // End any current speech detection
    if (_isCurrentlySpeaking) {
      _isCurrentlySpeaking = false;
      onSpeechActivity?.call(SpeechActivityEvent.ended);
    }

    // Reset counters
    _consecutiveSilentFrames = 0;
    _consecutiveVoiceFrames = 0;
  }

  /// Notify VAD that TTS has finished playing
  void notifyTTSDidFinish() {
    _isTTSActive = false;

    // Immediately restore threshold for instant response
    energyThreshold = _baseEnergyThreshold;

    _logger.info(
      '🔇 TTS finished - VAD threshold restored to ${energyThreshold.toStringAsFixed(6)}',
    );

    // Reset state for immediate readiness
    _recentEnergyValues.clear();
    _consecutiveSilentFrames = 0;
    _consecutiveVoiceFrames = 0;
    _isCurrentlySpeaking = false;

    // Prime the VAD to be ready for immediate detection
    _debugFrameCount = 0;
  }

  /// Set TTS threshold multiplier for feedback prevention
  void setTTSThresholdMultiplier(double multiplier) {
    _ttsThresholdMultiplier = math.max(2.0, math.min(5.0, multiplier));
    _logger.info('📝 TTS threshold multiplier set to ${_ttsThresholdMultiplier}x');
  }

  // MARK: - Debug Helpers

  /// Update debug statistics with new energy value
  void _updateDebugStatistics(double energy) {
    _recentEnergyValues.add(energy);
    if (_recentEnergyValues.length > _maxRecentValues) {
      _recentEnergyValues.removeAt(0);
    }
  }
}
